programs/stake/src/stake_state.rs

```diff
@@ -456,21 +457,39 @@ impl Stake {
     pub fn calculate_points_and_credits(
         &self,
         new_vote_state: &VoteState,
-        stake_history: Option<&StakeHistory>,
+        stake_context: Option<(Epoch, &StakeHistory)>,
         inflation_point_calc_tracer: &mut Option<impl FnMut(&InflationPointCalculationEvent)>,
     ) -> (u128, u64) {
         // if there is no newer credits since observed, return no point
+        let mut stake_at_target_epoch = None;
+        if let Some((target_epoch, stake_history)) = stake_context {
+            let stake = self.delegation.stake(target_epoch, Some(stake_history));
+            stake_at_target_epoch = Some(stake);
+
+            if let Some(inflation_point_calc_tracer) = inflation_point_calc_tracer {
+                inflation_point_calc_tracer(&InflationPointCalculationEvent::Stake(stake));
+            }
+        }
+
         if new_vote_state.credits() <= self.credits_observed {
             return (0, 0);
         }
 
         let mut points = 0;
         let mut new_credits_observed = self.credits_observed;
 
-        for (epoch, final_epoch_credits, initial_epoch_credits) in
-            new_vote_state.epoch_credits().iter().copied()
+        if let Some((epoch, final_epoch_credits, initial_epoch_credits)) =
+            new_vote_state.epoch_credits().last().copied()
         {
-            let stake = u128::from(self.delegation.stake(epoch, stake_history));
+            let stake = if let Some((target_epoch, _stake_history)) = stake_context {
+                if epoch != target_epoch {
+                    return (0, 0);
+                }
```

Comment:
write a test for this; this bug is later found after manual testing.

Reply:
well, if we not exclude historical vote_state's epoch credits, this way, we might accidentally reward stale delegations well past its (last and stale) epoch_credits.